### DISCUSSION TITLE
Link between electrode channels and units

### DISCUSSION THREAD

**Discussion by GaelleChapuis** [dandi/helpdesk#103] [Q&A]
Hello,

I am using this dataset for testing
```
dandiset_id = '000409'
filepath = 'sub-KS042/sub-KS042_ses-07dc4b76-5b93-4a03-82a0-b3d9cc73f412_behavior+ecephys+image.nwb'
```
And loading the electrode channels data like so:
```
channels_all = nwb.ec_electrodes.to_dataframe()
```
This session has recordings from 2 Neuropixels probes `probe00` and `probe01` (each have 384 channels), in AP / LF band. 
The channels dataframe contains thus 2 (AP / LF) x 2 (`probe00`/`probe01`) x 384 (channels).

I am loading the units data like so, and am filtering to keep those just for a single electrode:
```
units = nwb.units.to_dataframe()
probe_label = 'probe00'
units_one_probe = units[units.probe_name == probe_label]
```
Note here that the probe label is `probe00`.

I am now trying to retrieve the corresponding information in the channels table, i.e. I want to load the information for the AP channels, just for the corresponding `probe_label`.
I am thus looking for a column in the channels table that would contain the label `probe00` or `probe01`.

The column `channels.group_name` contains the unique values `'NeuropixelsShank00', 'NeuropixelsShank01'`. Can I safely assume that `probe00` is `'NeuropixelsShank00'` (it would be easier programmatically if the convention was to keep the same key name)? Or is there another place to retrieve this information from (either in channels, or in units)?

Thank you !

**Comment by CodyCBakerPhD** [comment]
I'd recommend checking out this NWB tutorial on [ecephys metadata](https://pynwb.readthedocs.io/en/stable/tutorials/domain/ecephys.html#extracellular-electrophysiology-data) to get a complete explanation of cross-linkage between the various data types

The short answer to the question is: the `electrodes` table has a column `groups` that links each electrode to a single `ElectrodeGroup` (used to model qualitative structural difference across a probe) which itself must link to a `Device` (which is lowest level representation of an individual probe in this case - that's also where the trajectory estimate and serial number will go in the next round of conversions).

If you haven't yet, I'd recommend checking out some of these NWB files using the NWB Widgets (`pip install nwbwidgets`), which visually lays out much of this file structure directly (and has a few neat plots such as PSTH from the `untis` tab)

```python
from nwbwidgets import nwb2widget

nwb2widget(nwbfile)  # using the same nwbfile you're currently accessing data from
```
